File: scripts/benchmark_pinyin_regex_phrase.py
#!/usr/bin/env python3
"""Benchmark Python pinyin_regex_phrase-compatible query construction."""

from __future__ import annotations

import argparse
import csv
import re
import statistics
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Callable


DEFAULT_QUERIES = (
    "lun",
    "lunlun",
    "zhengshuang",
    "wangchongyang",
    "xian",
    "wchy",
    "zh sh",
    "   ",
    "abc",
)


@dataclass(frozen=True)
class BenchmarkResult:
    label: str
    mode: str
    rows: int
    nonnull: int
    checksum: int
    best_ms: float
    median_ms: float
    mean_ms: float
    per_row_us: float
    last_result: str | None
    runs_ms: list[float]


def parse_args() -> argparse.Namespace:
    root_dir = Path(__file__).resolve().parents[1]
    parser = argparse.ArgumentParser(
        description="Benchmark Python pinyin_regex_phrase-compatible query construction."
    )
    parser.add_argument(
        "--rows",
        type=int,
        default=20000,
        help="number of query strings to build",
    )
    parser.add_argument(
        "--runs",
        type=int,
        default=5,
        help="number of timed runs",
    )
    parser.add_argument(
        "--token-file",
        default=str(root_dir / "sql/data/pinyin_token.csv"),
        help="pinyin token CSV used by the Python implementation",
    )
    parser.add_argument(
        "--queries",
        default=",".join(DEFAULT_QUERIES),
        help="comma-separated query mix",
    )
    parser.add_argument(
        "--mode",
        choices=("tokens", "query"),
        default="tokens",
        help="benchmark token pattern construction or pdb.query string construction",
    )
    return parser.parse_args()


def load_python_helper(token_file: str, mode: str) -> Callable[[str], list[str] | str | None]:
    tokens: list[str] = []
    with open(token_file, newline="") as fh:
        for token, category in csv.reader(fh):
            if category == "1" or token in {"zh", "ch", "sh"}:
                tokens.append(token)

    tokens = sorted(set(tokens), key=lambda item: (-len(item), item))
    tokenizer = re.compile("(" + "|".join(re.escape(token) for token in tokens) + "|[a-z])")
    all_english = re.compile(r"[a-zA-Z\s]+$")

    def helper(value: str) -> list[str] | str | None:
        if not all_english.match(value):
            return []
        pinyins = tokenizer.findall(value.lower())
        patterns = [f"{token}.*" for token in pinyins]
        if mode == "tokens":
            return patterns
        if not patterns:
            return "pdb.empty()"
        if len(patterns) < 2:
            return f"pdb.regex({patterns[0]!r})"
        return f"pdb.regex_phrase({patterns!r})"

    return helper


def build_inputs(rows: int, queries_text: str) -> list[str]:
    queries = [query for query in queries_text.split(",")]
    if rows < 0:
        raise ValueError("--rows must be non-negative")
    if not queries:
        raise ValueError("--queries must contain at least one query")
    return [queries[i % len(queries)] for i in range(rows)]


def run_benchmark(
    helper: Callable[[str], list[str] | str | None],
    label: str,
    mode: str,
    inputs: list[str],
    runs: int,
) -> BenchmarkResult:
    if runs <= 0:
        raise ValueError("--runs must be positive")

    for value in inputs[: min(1000, len(inputs))]:
        helper(value)

    durations: list[float] = []
    nonnull = 0
    checksum = 0
    last_result: list[str] | str | None = None

    for _ in range(runs):
        start = time.perf_counter()
        local_nonnull = 0
        local_checksum = 0
        for value in inputs:
            result = helper(value)
            if result is not None:
                local_nonnull += 1
                if isinstance(result, list):
                    local_checksum += len(result)
                    local_checksum += sum(len(item) for item in result)
                else:
                    local_checksum += len(result)
                last_result = result
        elapsed_ms = (time.perf_counter() - start) * 1000.0
        durations.append(elapsed_ms)
        nonnull = local_nonnull
        checksum = local_checksum

    best_ms = min(durations)
    return BenchmarkResult(
        label=label,
        mode=mode,
        rows=len(inputs),
        nonnull=nonnull,
        checksum=checksum,
        best_ms=best_ms,
        median_ms=statistics.median(durations),
        mean_ms=statistics.fmean(durations),
        per_row_us=(best_ms * 1000.0 / len(inputs)) if inputs else 0.0,
        last_result=repr(last_result),
        runs_ms=durations,
    )


def print_result(result: BenchmarkResult) -> None:
    print("")
    print("=== Query Builder Benchmark: Python pinyin_regex_phrase ===")
    print(f"helper: {result.label}")
    print(f"mode: {result.mode}")
    print("")
    print("| metric | value |")
    print("| --- | ---: |")
    print(f"| rows | {result.rows} |")
    print(f"| non-null queries | {result.nonnull} |")
    print(f"| best | {result.best_ms:.3f} ms |")
    print(f"| median | {result.median_ms:.3f} ms |")
    print(f"| mean | {result.mean_ms:.3f} ms |")
    print(f"| best per row | {result.per_row_us:.3f} us |")
    print(f"| checksum | {result.checksum} |")
    print("")
    print(f"runs_ms: {', '.join(f'{item:.3f}' for item in result.runs_ms)}")
    print(f"last_result: {result.last_result}")


def main() -> int:
    args = parse_args()
    inputs = build_inputs(args.rows, args.queries)
    helper = load_python_helper(args.token_file, args.mode)
    label = "pure Python compatible implementation"

    result = run_benchmark(helper, label, args.mode, inputs, args.runs)
    print_result(result)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/bin/benchmark_pinyin_regex_phrase.rs
use std::env;
use std::fs;
use std::hint::black_box;
use std::path::PathBuf;
use std::time::Instant;

use pg_pinyin::regex_phrase::{
    RegexTokenDictionary, pinyin_regex_phrase_patterns, tokens_from_pinyin_token_csv,
};

const DEFAULT_QUERIES: [&str; 9] = [
    "lun",
    "lunlun",
    "zhengshuang",
    "wangchongyang",
    "xian",
    "wchy",
    "zh sh",
    "   ",
    "abc",
];

struct Args {
    rows: usize,
    runs: usize,
    token_file: PathBuf,
    queries: Vec<String>,
    mode: Mode,
}

#[derive(Clone, Copy)]
enum Mode {
    Tokens,
    Query,
}

struct RunResult {
    elapsed_ms: f64,
    nonnull: usize,
    checksum: usize,
    last_result: Option<String>,
}

fn parse_args() -> Args {
    let root_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let mut args = Args {
        rows: 20000,
        runs: 5,
        token_file: root_dir.join("sql/data/pinyin_token.csv"),
        queries: DEFAULT_QUERIES
            .iter()
            .map(|query| query.to_string())
            .collect(),
        mode: Mode::Tokens,
    };

    let mut iter = env::args().skip(1);
    while let Some(arg) = iter.next() {
        match arg.as_str() {
            "--rows" => {
                args.rows = iter
                    .next()
                    .expect("--rows requires a value")
                    .parse()
                    .expect("--rows must be a non-negative integer");
            }
            "--runs" => {
                args.runs = iter
                    .next()
                    .expect("--runs requires a value")
                    .parse()
                    .expect("--runs must be a positive integer");
            }
            "--token-file" => {
                args.token_file =
                    PathBuf::from(iter.next().expect("--token-file requires a value"));
            }
            "--queries" => {
                args.queries = iter
                    .next()
                    .expect("--queries requires a value")
                    .split(',')
                    .map(str::to_string)
                    .collect();
            }
            "--mode" => {
                args.mode = match iter.next().expect("--mode requires a value").as_str() {
                    "tokens" => Mode::Tokens,
                    "query" => Mode::Query,
                    other => panic!("unsupported --mode {other}; expected tokens or query"),
                };
            }
            "--help" | "-h" => {
                println!(
                    "Usage: benchmark_pinyin_regex_phrase [--rows N] [--runs N] [--token-file PATH] [--queries CSV] [--mode tokens|query]"
                );
                std::process::exit(0);
            }
            other => panic!("unsupported argument: {other}"),
        }
    }

    if args.runs == 0 {
        panic!("--runs must be positive");
    }
    if args.queries.is_empty() {
        panic!("--queries must contain at least one query");
    }

    args
}

fn build_inputs(rows: usize, queries: &[String]) -> Vec<String> {
    (0..rows)
        .map(|idx| queries[idx % queries.len()].clone())
        .collect()
}

fn build_query_string(patterns: &[String]) -> String {
    if patterns.len() == 1 {
        format!("pdb.regex({:?})", patterns[0])
    } else {
        format!("pdb.regex_phrase({patterns:?})")
    }
}

fn run_once(inputs: &[String], dictionary: &RegexTokenDictionary, mode: Mode) -> RunResult {
    let start = Instant::now();
    let mut nonnull = 0usize;
    let mut checksum = 0usize;
    let mut last_result = None;

    for value in inputs {
        let result = pinyin_regex_phrase_patterns(value, false, dictionary);
        if let Some(patterns) = result {
            nonnull += 1;
            checksum += patterns.len();
            match mode {
                Mode::Tokens => {
                    checksum += patterns.iter().map(String::len).sum::<usize>();
                    last_result = Some(format!("{patterns:?}"));
                }
                Mode::Query => {
                    let query = build_query_string(&patterns);
                    checksum += query.len();
                    last_result = Some(query);
                }
            }
        }
    }

    RunResult {
        elapsed_ms: start.elapsed().as_secs_f64() * 1000.0,
        nonnull,
        checksum: black_box(checksum),
        last_result,
    }
}

fn main() {
    let args = parse_args();
    let token_csv = fs::read_to_string(&args.token_file).expect("failed to read token file");
    let dictionary = RegexTokenDictionary::from_tokens(tokens_from_pinyin_token_csv(&token_csv));
    let inputs = build_inputs(args.rows, &args.queries);

    for value in inputs.iter().take(1000) {
        let _ = pinyin_regex_phrase_patterns(value, false, &dictionary);
    }

    let mut runs = Vec::with_capacity(args.runs);
    for _ in 0..args.runs {
        runs.push(run_once(&inputs, &dictionary, args.mode));
    }

    let mut durations: Vec<f64> = runs.iter().map(|run| run.elapsed_ms).collect();
    durations.sort_by(f64::total_cmp);
    let best = durations[0];
    let median = durations[durations.len() / 2];
    let mean = durations.iter().sum::<f64>() / durations.len() as f64;
    let last = runs.last().expect("at least one run");
    let mode = match args.mode {
        Mode::Tokens => "tokens",
        Mode::Query => "query",
    };

    println!();
    println!("=== Query Builder Benchmark: Rust standalone pinyin_regex_phrase ===");
    println!("helper: pure Rust implementation");
    println!("mode: {mode}");
    println!("token_count: {}", dictionary.token_count());
    println!();
    println!("| metric | value |");
    println!("| --- | ---: |");
    println!("| rows | {} |", args.rows);
    println!("| non-null queries | {} |", last.nonnull);
    println!("| best | {best:.3} ms |");
    println!("| median | {median:.3} ms |");
    println!("| mean | {mean:.3} ms |");
    println!(
        "| best per row | {:.3} us |",
        best * 1000.0 / args.rows as f64
    );
    println!("| checksum | {} |", last.checksum);
    println!();
    println!(
        "runs_ms: {}",
        runs.iter()
            .map(|run| format!("{:.3}", run.elapsed_ms))
            .collect::<Vec<_>>()
            .join(", ")
    );
    println!(
        "last_result: {}",
        last.last_result.as_deref().unwrap_or("<null>")
    );
}
